crates/evm/src/evm/system_contracts/src/Bridge.sol

```diff
@@ -165,7 +185,76 @@ contract Bridge is Ownable2StepUpgradeable {
         operator = _operator;
         emit OperatorUpdated(operator, _operator);
     }
-    
+
+    /// @notice Stores the filler of a certain withdrawal after the a user's withdrawal is covered on Bitcoin side
+    /// @param withdrawTp Transaction parameters of the withdrawal transaction on Bitcoin
+    /// @param inputIndex Index of the input that is the withdrawal UTXO (withdrawing user's ANYONECANPAY)
+    /// @param withdrawId ID of the withdrawal action
+    function declareWithdrawFiller(TransactionParams calldata withdrawTp, uint256 inputIndex, uint256 withdrawId) external {
+        validateAndCheckInclusion(withdrawTp);
+        bytes memory input = BTCUtils.extractInputAtIndex(withdrawTp.vin, inputIndex);
+        bytes32 txId = BTCUtils.extractInputTxIdLE(input);
+        bytes4 index = BTCUtils.extractTxIndexLE(input);
+        UTXO memory utxo = withdrawalUTXOs[withdrawId];
+        require(utxo.txId == txId && utxo.outputId == index, "not matching UTXO");
+
+        (, uint256 nOuts) = BTCUtils.parseVarInt(withdrawTp.vout);
+        bytes memory output = BTCUtils.extractOutputAtIndex(withdrawTp.vout, nOuts - 1);
+        bytes memory opReturnData = BTCUtils.extractOpReturnData(output);
+        uint256 withdrawFillerId = uint256(bytesToBytes32(opReturnData));
+        withdrawFillers[withdrawId] = getInternalOperatorId(withdrawFillerId);
+        emit WithdrawFillerDeclared(withdrawId, withdrawFillerId);
+    }
+
+    /// @notice Marks an operator as malicious if the operator burned their slashOrTake transaction as if they filled a withdrawal even though they didn't fill a withdrawal
+    /// @param slashOrTakeTp Transaction parameters of the slashOrTake transaction on Bitcoin
+    function markMaliciousOperator(TransactionParams calldata slashOrTakeTp) external {
+        validateAndCheckInclusion(slashOrTakeTp);
+        
+        (, uint256 nOuts) = BTCUtils.parseVarInt(slashOrTakeTp.vout);
+        bytes memory output = BTCUtils.extractOutputAtIndex(slashOrTakeTp.vout, nOuts - 1);
+        bytes memory opReturnData = BTCUtils.extractOpReturnData(output);
+        bytes32 moveTxId = bytesToBytes32(opReturnData.slice(0, 32));
+        uint256 operatorId = uint256(bytesToBytes32(opReturnData.slice(32, opReturnData.length - 32)));
+        uint256 depositId = txIdToDepositId[moveTxId];
```

Comment:
We should raise error in here if txIdToDepositId[moveTxId] is empty.
And also there is a mismatch between depositId and withdrawalUTXOs, depositId starts from 1 but withdrawalUTXOs starts from 0